samcli/lib/deploy/deployer.py

```diff
@@ -409,6 +416,8 @@ def describe_stack_events(self, stack_name, time_stamp_marker, **kwargs):
                 if self._check_stack_not_in_progress(stack_status):
                     stack_change_in_progress = False
                     break
+                # Reset retry attempts if iteration is a success to use client_sleep again
+                retry_attempts = 0
```

Comment:
If found a bug at https://github.com/aws/aws-sam-cli/blob/d1db668924ba74b0f71813da18591de1b80b4142/samcli/lib/deploy/deployer.py#L371-L372

Which causes to ignore the `client_sleep` and always send requests without delay if there was one exception in previous iterations, even if after the retry all the other iterations are successful.

With this change, the function stops only if there are consecutive retries exceptions.

But we could also consider the retry attemps globally no matter if there was successes in between.
In that case I could add a condition on the sleep to skip only if the last iteration was in exception.

Let me know what's best @moelasmar @jfuss

Reply:
I think it is better to stop only if there consecutive retries exceptions, this increases the probability that the customer can finish the `sam deploy` command successfully.

I have a question why we skip the sleep time after some exception, I think waiting some times help to resolve some exceptions like the throttling exceptions, and this will increase the command success probability. May be the cons of this approach is the command will take a longer time to finish (either success or fail).

Reply:
I think that is ok. It might not be ideal in all cases but we can always adjust this in the future (I would think). Let me ponder this a little more to make sure that all seems reasonable. 

@BourgoisMickael Sorry I missed the update on the PR.